# 에러 바운더리의 특징

단, React의 Error Boundary는 본래 이벤트 핸들러의 에러를 포착하지 않는다고 한다. 이 경우, 항상 try/catch를 거치도록해서 catch에서 잡힌 error를 throw 해주면 잡을 수 있다고 한다. react-query는 `useErrorboundary` 옵션을 제공하고 있고, 이걸 사용해야만 에러바운더리가 API 에러를 캐치할 수 있다.

## 에러바운더리가 캐치하지 못하는 에러는?

에러바운더리가 캐치하지 못하는 에러에 대해서 설명(+실행컨텍스트, 콜스택) <https://happysisyphe.tistory.com/66>

```tsx
// NOTE 이렇게 구현하면 실행컨텍스트의 콜 스택에서 timeout이 먼저 popout되어서 에러 바운더리까지 에러가 전달되지 않음.
const DelayedErrorComponent: React.FC = () => {
  useEffect(() => {
    const timer = setTimeout(() => {
      throw new Error("Rendering took too long");
    }, 10000); // 10초 지연

    return () => clearTimeout(timer);
  }, []);

  return <div>Delayed Error Component</div>;
};

export default DelayedErrorComponent;
```

왜 안되는지 그 이유

```txt
이 코드에서는 useEffect 훅 내부에서 발생한 에러는 React의 에러 경계를 벗어나기 때문에 에러 바운더리에서 캐치되지 않습니다.

useEffect 내부에서 발생한 에러는 해당 컴포넌트의 렌더링 주기에서 발생하지 않고, 비동기적으로 발생합니다. 이러한 비동기적인 코드에서 발생한 에러는 React의 컴포넌트 트리를 벗어나기 때문에 에러 바운더리에서 캐치되지 않습니다.

따라서 useEffect 내부에서 에러를 발생시킨 경우, 이를 캐치하려면 useEffect 내부에서 직접 에러 처리를 해야 합니다. 일반적으로 try-catch 블록을 사용하여 에러를 처리하고, 그에 따라 컴포넌트의 상태나 UI를 업데이트할 수 있습니다.

에러 바운더리는 주로 렌더링 중에 발생한 에러를 캐치하고 처리하는 데 사용되며, 이러한 경우에는 컴포넌트의 렌더링 주기 내에서 발생한 에러를 캐치할 수 있습니다.
```

이렇게 해야 에러 바운더리가 에러를 캐치함

```tsx
export default function TestSuspenseFallback() {
  // NOTE 10초 이상 해당 컴포넌트가 렌더링되면 에러 발생시키기.
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    const timer = setTimeout(() => {
      setIsError(true);
    }, 10000); // 10초 지연

    return () => clearTimeout(timer);
  }, []);

  if (isError) {
    throw new Error("Rendering took too long");
  }

  return (
    <ErrorBoundary fallback={<DequeueEmpty />}>
      <div>
        Now is Suspense
        <Button
          onClick={() => {
            throw new Error("Rendering took too long");
          }}
        >
          Dequeue Empty haha
        </Button>
      </div>
    </ErrorBoundary>
  );
}
```

- 상태 관리: 먼저, useState 훅을 사용하여 isError라는 상태를 관리합니다. 이 상태는 컴포넌트가 에러 상태에 있는지 여부를 나타냅니다.
- 렌더링 지연: useEffect 훅을 사용하여 컴포넌트가 마운트될 때 10초 타이머를 설정합니다. 이 타이머는 10초 후에 setIsError(true)를 호출하여 isError 상태를 true로 변경합니다.
- 에러 처리: 컴포넌트의 렌더 함수에서 isError 상태를 확인하고, 만약 true라면 throw new Error()를 호출하여 에러를 발생시킵니다.
- 에러 바운더리: 컴포넌트가 렌더링되면서 발생한 에러는 그 부모 컴포넌트나 최상위 컴포넌트의 에러 바운더리로 전파됩니다. 이러한 에러 바운더리는 에러가 발생한 컴포넌트의 트리에서 가장 가까운 위치에서 에러를 캐치합니다.
- 에러 캐치: 에러 바운더리에서는 try-catch 블록을 사용하여 자식 컴포넌트의 렌더링 중 발생한 에러를 캐치합니다. 이 경우, 해당 에러를 콘솔에 출력하거나 사용자에게 오류 메시지를 보여줄 수 있습니다.

읽어봄직한 컬럼: <https://velog.io/@mmmdo21/%EC%84%A0%EC%96%B8%EC%A0%81%EC%9C%BC%EB%A1%9C-%EC%97%90%EB%9F%AC%EC%B2%98%EB%A6%AC%ED%95%98%EA%B8%B0react-error-boundary>
